### DISCUSSION TITLE
请问提交上屏的API是否一定要在主线程执行，是否可以开放一个API让业务自行传入自定义的渲染Queue?

### DISCUSSION THREAD

**Discussion by edisongz** [Tencent/libpag#499] [Q&A] [answered]
目前业务有循环播放PAG的场景，Instrument Time Profile看到的瓶颈在libpag.
https://github.com/Tencent/libpag/blob/0d218296d1a813fe571b9a1bad255588770661c7/src/platform/ios/private/GPUDrawable.mm#L154

<img width="512" alt="image" src="https://user-images.githubusercontent.com/3645329/187202241-1f4a36ee-ec8f-429f-bc53-6638f9e5dd96.png">

**Comment by kevingpqi123** [comment, marked as answer]
上屏是需要在主线程
如果遇到性能瓶颈：首先需要看下素材的优化空间，相同的动画不同的实现方式性能差异特别大，让你们的设计师看下这篇文档：
https://pag.art/docs/optimization.html

另外，pag 是支持离屏渲染的，通过PAGPlayer、PAGSurface的相关接口，支持自由控制渲染、播放逻辑

**Reply by domchen** [reply]
你可以看下源码，从CALayer创建的GPUDrawable才会post到主线程上屏，这是iOS接口的限制，CALayer只能在主线程presentBuffer。如果你是从一个离屏对象创建的，比如CVPixelBuffer，那么是在调用线程的。

**Reply by edisongz** [reply]
<img width="591" alt="image" src="https://user-images.githubusercontent.com/3645329/187901982-dbc2397b-c399-45b6-ac34-4e301a17de51.png">
如图，官方demo可以独立线程正常渲染PAG帧。

UIKit必须主线程执行可以理解，Core Animation和OpenGL ES应该都是可以自定义渲染线程执行才对，播放器或游戏引擎之类的渲染都是独立线程；否则场景稍微复杂点，PAGView就会卡顿。

**Reply by domchen** [reply]
> <img alt="image" width="591" src="https://user-images.githubusercontent.com/3645329/187901982-dbc2397b-c399-45b6-ac34-4e301a17de51.png">
> 
> 如图，官方demo可以独立线程正常渲染PAG帧。
> UIKit必须主线程执行可以理解，Core Animation和OpenGL ES应该都是可以自定义渲染线程执行才对，播放器或游戏引擎之类的渲染都是独立线程；否则场景稍微复杂点，PAGView就会卡顿。

你好像没有看明白这段代码的逻辑。走到这里的时候已经执行完渲染了，OpenGL的渲染是在异步线程独立跑的，但是最后有一个刷新buffer的操作是必须在主线程执行，任何要上屏的OpenGL都需要这样操作，就算游戏也是一样。这个present只是把opengl渲染的结果刷新给layer。

**Reply by domchen** [reply]
https://stackoverflow.com/questions/34922117/is-it-safe-to-call-eaglcontext-presentrenderbuffer-on-a-secondary-thread 这里有个帖子正好在说是否可以在子线程present的，总结应该是早期的版本是有问题的，但是后来不知道从哪个版本开始已经没有问题了。我们会继续调研一下看看是否还有必要post到主线程进行present。但是现在只是present这一个小操作是在主线程执行的，真正的每帧渲染过程还是异步的。
